engine: move protos to (new) enginepb package

tbg opened the pull request:
```
Under leader-evaluated Raft, the MVCCStats delta will be computed before the
proposal. Since the delta only commits when the write set does, and since
updating the protobufs is sensitive to ordering (unlike using, for example,
increment operations on individual keys), the delta will be sent along with the
Raft command.

This necessitates adding a `MVCCStats` object to `roachpb.RaftCommand`, but the
references from `storage/engine` to `roachpb` in methods which sit on engine
protos make this circular.

This change shells out the engine protos into their own package instead.
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7166)

<!-- Reviewable:end -->

tbg committed:
engine: move pb receiver in own file

No code changes.

tbg committed:
engine: remove dependency on roachpb from protos

Under leader-evaluated Raft, the MVCCStats delta will be computed before the
proposal. Since the delta only commits when the write set does, and since
updating the protobufs is sensitive to ordering (unlike using, for example,
increment operations on individual keys), the delta will be sent along with the
Raft command.

This necessitates adding a `MVCCStats` object to `roachpb.RaftCommand`, but the
references from `storage/engine` to `roachpb` in methods which sit on engine
protos make this circular.

This commit moves these methods off the receiver in preparation for splitting
out the protos from `storage/engine` into their own package which does not
reference `roachpb`.

tbg committed:
engine: move protos to `enginepb` package

This new package has no dependency on `roachpb` and thus allows usage of
MVCCStats from `roachpb`.

tbg committed:
enginepb: rename mvcc{pb->}.go

tschottdorf commented:
Ah. I overlooked that (of course) `MVCCMetadata` makes use of `roachpb.{Timestamp,TxnMeta}` which means that the dependency isn't actually resolved with this PR.

In the interest of separating the primary on-disk protos and `roachpb`, an immediate way to solve this moving `TxnMeta` into `enginepb`. `Timestamp` might be best kept inside `util/hlc` (though of course it will also be on-disk).

I'll see what that looks like, but would like to hear other ideas.

tamird commented on storage/engine/mvcc.go:
```diff
@@ -49,6 +50,36 @@ var (
+// the values that it records when the structure is initially stored. Specifically,
+// MVCCStats is stored on the RangeStats key, which means that its creation will
+// have an impact on system-local data size and key count.
+func AccountForSelf(ms *enginepb.MVCCStats, rangeID roachpb.RangeID) error {
```
can't this go back to being a method now?

tamird replied:
ah this method still uses roachpb. got it.

tschottdorf replied:
Yeah, awkward. Not sure how to make this look better.

On Sat, Jun 11, 2016 at 4:21 PM Tamir Duberstein <notifications@github.com>
wrote:

> In storage/engine/mvcc.go
> <https://github.com/cockroachdb/cockroach/pull/7166#discussion_r66711224>:
>
> > @@ -49,6 +50,36 @@ var (
> >  	NilKey = MVCCKey{}
> >  )
> >
> > +// AccountForSelf adjusts ms to account for the predicted impact it will have on
> > +// the values that it records when the structure is initially stored. Specifically,
> > +// MVCCStats is stored on the RangeStats key, which means that its creation will
> > +// have an impact on system-local data size and key count.
> > +func AccountForSelf(ms *enginepb.MVCCStats, rangeID roachpb.RangeID) error {
>
> can't this go back to being a method now?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/7166/files/c986fa5e308346a0050cf4d2e76e02ccfdeb7527..a0846d3e36776522759174b0bff8589c0f900a82#r66711224>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AE135F-v-Io0DeTJJCuLnD--oJSPlX-Hks5qKxixgaJpZM4IzeMe>
> .
>
-- 

-- Tobias

tschottdorf commented:
I added more commits moving `Timestamp` to `hlc` and `TxnMeta` into `enginepb`. Not convinced this is the way to go (in particular see the last commit message).

tschottdorf commented:
Freshly rebased and ready for review. Apologies for the diffs - r8-r14 are the current ones (only rebased).

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

It would have also worked to move (some of) the engine protos back into roachpb, right? I'm not suggesting that now (now that the work is done it's a good change), but I think that would have also been an acceptable answer. 

---

Review status: 0 of 125 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 279 \[r14\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7166#-KKCF8lI5U8m5OW5T6xA:-KKCF8lI5U8m5OW5T6xB:830741389) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a2b4866881db3b8419e8b580378caa02701787/storage/replica.go#L279)):*
> ```Go
> 		// gcThreshold is the GC threshold of the replica. Reads and writes must
> 		// not happen <= this time.
> 		gcThreshold hlc.Timestamp
> ```

Why do we have two of these fields? (bad merge?)

---

*[storage/engine/enginepb/mvcc.proto, line 25 \[r14\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7166#-KKCGVOp61vL3EVaG_dT:-KKCGVOp61vL3EVaG_dU:1356883486) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a2b4866881db3b8419e8b580378caa02701787/storage/engine/enginepb/mvcc.proto#L25)):*
> ```protobuf
> import weak "gogoproto/gogo.proto";
> 
> // TODO(tschottdorf): Should not live in enginepb (but can't live in roachpb
> ```

Hmm. It's most closely associated with `kv`, but a dependency chain of `roachpb -> enginepb -> kvpb` just feels wrong.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7166#-:-KKCGzZKUtjmErM4uzzE:-1346039627)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 46 files at r3, 1 of 2 files at r4, 125 of 131 files at r8, 8 of 10 files at r9, 45 of 46 files at r10, 2 of 2 files at r11, 102 of 102 files at r12, 37 of 37 files at r13, 11 of 11 files at r14.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[storage/engine/engine_test.go, line 247 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7166#-KKEFvRVfwli6LBhgVdD:-KKEFvRVfwli6LBhgVdE:-673792339) ([raw file](https://github.com/cockroachdb/cockroach/blob/66f16819d73f99495ea3819d3ff67a55fec430b0/storage/engine/engine_test.go#L247)):*
> ```Go
> 				return nil
> 			}
> 			valueBytes, err := MakeValue(*m).GetBytes()
> ```

nit: maybe make `m` a value =/

---

*[storage/engine/engine_test.go, line 302 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7166#-KKEFyx-kn1oAjTpsMYQ:-KKEFyx-kn1oAjTpsMYR:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/66f16819d73f99495ea3819d3ff67a55fec430b0/storage/engine/engine_test.go#L302)):*
> ```Go
> 					t.Fatal(err)
> 				}
> 				valueBytes, err := MakeValue(*m).GetBytes()
> ```

ditto

---

*[storage/engine/enginepb/mvcc.proto, line 45 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7166#-KKEJ76JxbS1AD3e9lO1:-KKEJ76JxbS1AD3e9lO2:1544755195) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db69b2692853dcb50bb2b34c977397264fc5a47/storage/engine/enginepb/mvcc.proto#L45)):*
> ```protobuf
>   // the first key read or written during the transaction and determines which
>   // range in the cluster will hold the transaction record.
>   optional bytes key = 3; // TODO(tschottdorf): [(gogoproto.casttype) = "Key"];
> ```

bummer. can you file an issue about moving roachpb.Key to keys?

---

*[storage/engine/enginepb/mvcc.proto, line 166 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7166#-KKEJAZRHZ5tVJ1vGgS9:-KKEJAZRHZ5tVJ1vGgSA:-1213019852) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db69b2692853dcb50bb2b34c977397264fc5a47/storage/engine/enginepb/mvcc.proto#L166)):*
> ```protobuf
>   optional sfixed64 sys_count = 13 [(gogoproto.nullable) = false];
> }
> ```

more random whitespace

---

*[util/hlc/timestamp.proto, line 15 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7166#-KKEIVWUtjTK1leCH8zG:-KKEIVWUtjTK1leCH8zH:-64557983) ([raw file](https://github.com/cockroachdb/cockroach/blob/90b37a8d3c602dd3d36e1f22a7d0f378af7d0dc6/util/hlc/timestamp.proto#L15)):*
> ```protobuf
> // permissions and limitations under the License.
> //
> // Author: Jiajia Han (hanjia18@gmail.com)
> ```

might be more appropriate to put your own name here

---

*[util/hlc/timestamp.proto, line 38 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7166#-KKEITCL8hoZzdKROhQs:-KKEITCL8hoZzdKROhQt:-1019282229) ([raw file](https://github.com/cockroachdb/cockroach/blob/90b37a8d3c602dd3d36e1f22a7d0f378af7d0dc6/util/hlc/timestamp.proto#L38)):*
> ```protobuf
>   optional int32 logical = 2 [(gogoproto.nullable) = false];
> }
> ```

random trailing newlines

---

*[util/hlc/timestamp_test.go, line 15 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7166#-KKEIaWb5nQSJhWSuMq0:-KKEIaWb5nQSJhWSuMq1:1198873996) ([raw file](https://github.com/cockroachdb/cockroach/blob/90b37a8d3c602dd3d36e1f22a7d0f378af7d0dc6/util/hlc/timestamp_test.go#L15)):*
> ```Go
> // permissions and limitations under the License.
> //
> // Author: Spencer Kimball (spencer.kimball@gmail.com)
> ```

ditto probably more likely you

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7166)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Yep, that should also have worked (though I find roachpb big and random enough already).
Comments addressed.

---

Review status: 0 of 129 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 279 \[r14\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7166#-KKCF8lI5U8m5OW5T6xA:-KKEyT4uFQJCoU_t6H_L:807718035) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a2b4866881db3b8419e8b580378caa02701787/storage/replica.go#L279)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Why do we have two of these fields? (bad merge?)
</details>
Exactly. Fixed.

---

*[storage/engine/engine_test.go, line 247 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7166#-KKEFvRVfwli6LBhgVdD:-KKEz-QBeLGRuX-hXNs1:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/66f16819d73f99495ea3819d3ff67a55fec430b0/storage/engine/engine_test.go#L247)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: maybe make `m` a value =/
</details>
Done.

---

*[storage/engine/engine_test.go, line 302 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7166#-KKEFyx-kn1oAjTpsMYQ:-KKEz0mx7yZxdNy1GYMz:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/66f16819d73f99495ea3819d3ff67a55fec430b0/storage/engine/engine_test.go#L302)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ditto
</details>
Done.

---

*[storage/engine/enginepb/mvcc.proto, line 45 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7166#-KKEJ76JxbS1AD3e9lO1:-KKEzRNyUKJGe1Lfd90A:397045135) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db69b2692853dcb50bb2b34c977397264fc5a47/storage/engine/enginepb/mvcc.proto#L45)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> bummer. can you file an issue about moving roachpb.Key to keys?
</details>
Done: #7220

---

*[storage/engine/enginepb/mvcc.proto, line 166 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7166#-KKEJAZRHZ5tVJ1vGgS9:-KKEzg_0gKH2smTIN1i7:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/3db69b2692853dcb50bb2b34c977397264fc5a47/storage/engine/enginepb/mvcc.proto#L166)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> more random whitespace
</details>
Done.

---

*[storage/engine/enginepb/mvcc.proto, line 25 \[r14\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7166#-KKCGVOp61vL3EVaG_dT:-KKEzAq0B5vwYR3inUw1:-598250012) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6a2b4866881db3b8419e8b580378caa02701787/storage/engine/enginepb/mvcc.proto#L25)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Hmm. It's most closely associated with `kv`, but a dependency chain of `roachpb -> enginepb -> kvpb` just feels wrong.
</details>
I agree. Could always move this into a top-level thing (which could also hold TxnMeta).

---

*[util/hlc/timestamp.proto, line 15 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7166#-KKEIVWUtjTK1leCH8zG:-KKEzlm9y37A2SMrC0Et:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/90b37a8d3c602dd3d36e1f22a7d0f378af7d0dc6/util/hlc/timestamp.proto#L15)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> might be more appropriate to put your own name here
</details>
Done.

---

*[util/hlc/timestamp.proto, line 38 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7166#-KKEITCL8hoZzdKROhQs:-KKEzvxOxoAOCkMQ7yJc:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/90b37a8d3c602dd3d36e1f22a7d0f378af7d0dc6/util/hlc/timestamp.proto#L38)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> random trailing newlines
</details>
Done.

---

*[util/hlc/timestamp_test.go, line 15 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7166#-KKEIaWb5nQSJhWSuMq0:-KKEzzPSDQ6SV534oN1g:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/90b37a8d3c602dd3d36e1f22a7d0f378af7d0dc6/util/hlc/timestamp_test.go#L15)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ditto probably more likely you
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7166#-:-KKF-4aeV59j7DrRNNU7:-2116926096)*
<!-- Sent from Reviewable.io -->

tbg committed:
roachpb: move TxnMeta to enginepb

This removes the last dependency `enginepb`->`roachpb` but introduces some
weirdness:

- `IsolationType` necessarily moved to `enginepb` as well, so there are random
  references to `enginepb` throughout the codebase, which is not ideal.
- `TxnMeta.Key` is now `[]byte` (not `roachpb.Key`), which had surprisingly
  little fallout, but is nonetheless not idiomatic.

tbg committed:
roachpb: move Timestamp to util/hlc

In preparation for removing further dependencies of `enginepb` on `roachpb`.
